File: ds/schemas/plot.py
from pydantic import Field, BaseModel
import uuid

from datetime import datetime
from numpy import random
import enum


colors = ("RED", "BLUE", "GREEN")
_random_names = [
    "maggot",
    "varam",
    "kammus",
    "sarayn",
    "enarvyne",
    "alammu",
    "irethi",
    "neldam",
    "dren",
    "anasour",
    "irarvy",
    "vandal",
    "tadaves",
    "seran",
    "llaalam",
    "worker",
    "dalamus",
    "vandal",
    "gidave",
    "sendal",
    "othralen",
    "tedril",
    "girothran",
    "ararvy",
    "maryon",
    "llaala",
    "faveran",
    "gadaves",
    "uradasou",
    "berendal",
    "maggot",
    "heloth",
    "neldammu",
    "othren",
    "midaves",
    "deras",
    "vedaves",
    "ienevala",
]


class Plot(BaseModel):
    title: str | None = None
    ylabel: str | None = None
    xlabel: str | None = None
    figsize: tuple[int, int] = (20, 13)


class Point(BaseModel):
    id_: uuid.UUID = Field(default_factory=uuid.uuid4, alias="id")
    color: str

    lat: float
    lon: float


class Points(Plot):
    points: list[Point] = Field(
        example=[
            Point(
                color=random.choice(colors),
                lat=random.rand() * 20 + 40,
                lon=random.rand() * 20 + 40,
            )
            for _ in range(random.randint(10, 20))
        ]
    )


class Bar(Plot):
    values: dict[str, float] = Field(
        example={
            random.choice(_random_names): random.randint(0, 100)
            for _ in range(random.randint(10, 20))
        }
    )

    limit: int | None = Field(None, example=None)
    asc: bool = True


class TimeSeries(Plot):
    values: dict[str, dict[datetime, float]] = Field(
        example={
            random.choice(_random_names): {
                datetime(
                    2024, random.randint(1, 12), random.randint(1, 28)
                ): random.randint(0, 100)
                for _ in range(20)
            }
            for _ in range(3)
        }
    )

    only_time: bool = False


class GraphEdge(BaseModel):
    first: str
    second: str
    weight: float


class GraphNode(BaseModel):
    image: bytes | None = None
    weight: float = 1


class GraphLayout(str, enum.Enum):
    SPRING_LAYOUT = "spring"
    CIRCULAR_LAYOUT = "circular"
    SPECTRAL_LAYOUT = "spectral"
    CIRCULAR_TREE_LAYOUT = "circular_tree"


class Graph(Plot):
    edges: list[GraphEdge] = Field(
        example=[
            GraphEdge(
                first=random.choice(_random_names),
                second=random.choice(_random_names),
                weight=random.randint(0, 100),
            )
            for _ in range(30)
        ]
    )
    weighted_edges: bool = True
    layout: GraphLayout = GraphLayout.CIRCULAR_LAYOUT
    nodes: dict[str, GraphNode] | None = None
    root_node: str | None = None

File: core/supplier/ds_supplier/plot.go
package ds_supplier

import (
	"context"

	"github.com/pkg/errors"
)

type DrawInput struct {
	Title         string `json:"title,omitempty"`
	XLabel        string `json:"xlabel,omitempty"`
	YLabel        string `json:"ylabel,omitempty"`
	FigSize       []int  `json:"figsize,omitempty"`
	ImageFormat   string `json:"image_format,omitempty"`
	LabelFontSize int    `json:"label_font_size,omitempty"`
}

func (r *DrawInput) setDefault() {
	if r.LabelFontSize == 0 {
		r.LabelFontSize = 14
	}
}

type DrawBarInput struct {
	DrawInput

	Values map[string]float64 `json:"values"`
	Limit  int                `json:"limit,omitempty"`
	Asc    bool               `json:"asc"`
}

func (r *Supplier) DrawBar(ctx context.Context, input *DrawBarInput) ([]byte, error) {
	input.setDefault()

	body, err := r.sendRequest(ctx, "plot/histogram", input)
	if err != nil {
		return nil, errors.Wrap(err, "failed to draw")
	}

	return body, nil
}

type DrawTimeseriesInput struct {
	DrawInput

	Values   map[string]map[string]float64 `json:"values"`
	OnlyTime bool                          `json:"only_time"`
}

func (r *Supplier) DrawTimeseries(ctx context.Context, input *DrawTimeseriesInput) ([]byte, error) {
	input.setDefault()

	body, err := r.sendRequest(ctx, "plot/timeseries", input)
	if err != nil {
		return nil, errors.Wrap(err, "failed to draw")
	}

	return body, nil
}

type GraphEdge struct {
	First  string  `json:"first"`
	Second string  `json:"second"`
	Weight float64 `json:"weight"`
}

type GraphNode struct {
	Image  []byte  `json:"image,omitempty"`
	Weight float64 `json:"weight,omitempty"`
}

type DrawGraphInput struct {
	DrawInput

	Edges         []GraphEdge          `json:"edges,omitempty"`
	Layout        string               `json:"layout,omitempty"`
	WeightedEdges bool                 `json:"weighted_edges"`
	Nodes         map[string]GraphNode `json:"nodes,omitempty"`
	RootNode      string               `json:"root_node,omitempty"`
}

func (r *Supplier) DrawGraph(ctx context.Context, input *DrawGraphInput) ([]byte, error) {
	input.setDefault()

	body, err := r.sendRequest(ctx, "plot/graph", input)
	if err != nil {
		return nil, errors.Wrap(err, "failed to draw")
	}

	return body, nil
}

func (r *Supplier) DrawGraphAsHeatpmap(ctx context.Context, input *DrawGraphInput) ([]byte, error) {
	input.setDefault()

	body, err := r.sendRequest(ctx, "plot/graph-as-heatmap", input)
	if err != nil {
		return nil, errors.Wrap(err, "failed to draw")
	}

	return body, nil
}
